### PR TITLE
feat(config): allow hostname to be overridden via env var

### LINKED ISSUE #14819: Duplicate `get_hostname` implementations
We have [one implementation](https://github.com/vectordotdev/vector/blob/405b333869d0053cf9ee214196e8d76be38a82e8/src/lib.rs#L146-L148) of `get_hostname` for Vector itself, and a [separate one](https://github.com/vectordotdev/vector/blob/405b333869d0053cf9ee214196e8d76be38a82e8/lib/vrl/stdlib/src/get_hostname.rs#L3-L8) for the VRL stdlib. It's not a huge deal since they're both quite small and currently the same, but when we introduce changes like allowing hostname to be overridden by an env var we don't want them to get out of sync.

### PR DESCRIPTION
**Pull request by lukesteensen**
Resolves #14819.

### CODE REVIEW

**Comment by jszwedko** [comment]
I rebased this on `master`. @fuchsnj do you want to give this a look? I think it is probably moving in the opposite direction than you are going with reducing the coupling so curious what you think about sharing the `get_hostname` implementation across Vector and VRL. Maybe it is something that needs to be passed _into_ VRL?

File: lib/vector-common/src/lib.rs [hunk 1]
```diff
@@ -63,3 +63,19 @@ pub type Error = Box<dyn std::error::Error + Send + Sync + 'static>;
 /// Vector's basic result type, defined in terms of [`Error`] and generic over
 /// `T`.
 pub type Result<T> = std::result::Result<T, Error>;
+
+/// Retrieve the current hostname.
```
Comment (neuronull) [inline lib/vector-common/src/lib.rs:67]: Seems like the comment could indicate that it is retrieved from the env var if present otherwise is computed.

**Review by neuronull — COMMENTED** [review]
LGTM

**Comment by spencergilbert** [comment]
@jszwedko do we want to revive this and get it completed?

**Comment by jszwedko** [comment]
I'd say yes since we still have two implementations which could get out of sync. I think we'll need to take a different approach now that VRL is split out though where Vector will need to inject a function into the VRL runtime to do the host resolution rather than just referencing the same implementation. Given that, I'll close this since we still have the tracking issue.

**Pull request closed without merging**